# DingK_NanJ_et_al_tRNA-QTL

## Prerequisites

To run this project, you need the following prerequisites:

   Python 3.7.16
   
   Perl 5
   
   R 4.2.3
## 1. smRNA-seq process

   Code for small RNA-seq data processing workflow. 
   
## 2. tRNA quantification

   Code for tRNA quantification and normalization. 

## 3. tQTL calling

   Code for tRNA normalization, confounding factor (PEER factor) detection and tQTL calling.

## 4. Colocalization and MR analysis

  Code for colocalization and Mendelian Randomization analysis between different mQTLs (molecular QTLs) or mQTL and disease GWAS.

## 5. downstream analysis

**5.1 validation:** Code for comparing tRNA quantification results with sequencing methods specifically designed for tRNA, including ARM-seq and DM-tRNA-seq.

**5.2 codon dependent correlation:** Code for analyzing between codon demand and its supply.

**5.3 g-tRNA classification:** Code for tRNA classification based on genetics evidence.

**5.4 tQTL features:** Code for comparison of tQTLs across tissues and ancestries, tRNA location enrichment, tQTL regulator identification and differential expression analysis of knock-down experiment.

**5.5 integrative tQTL-pQTL analysis:** Code for the investigation of codon demand vs. regulon size, tRNA regulatory preferences for proteins, translational effect size balance, and GO enrichment analysis.
